python/cudf/cudf/core/udf/utils.py

```diff
@@ -0,0 +1,215 @@
+from typing import Callable
+
+import cachetools
+import numpy as np
+from numba import cuda, typeof
+from numba.np import numpy_support
+from numba.types import Poison, Tuple, boolean, int64, void
+from nvtx import annotate
+
+from cudf.core.dtypes import CategoricalDtype
+from cudf.core.udf.typing import MaskedType
+from cudf.utils import cudautils
+from cudf.utils.dtypes import (
+    BOOL_TYPES,
+    DATETIME_TYPES,
+    NUMERIC_TYPES,
+    TIMEDELTA_TYPES,
+)
+
+JIT_SUPPORTED_TYPES = (
+    NUMERIC_TYPES | BOOL_TYPES | DATETIME_TYPES | TIMEDELTA_TYPES
+)
+
+libcudf_bitmask_type = numpy_support.from_dtype(np.dtype("int32"))
+MASK_BITSIZE = np.dtype("int32").itemsize * 8
+
+precompiled: cachetools.LRUCache = cachetools.LRUCache(maxsize=32)
+
+
+@annotate("NUMBA JIT", color="green", domain="cudf_python")
+def _get_udf_return_type(argty, func: Callable, args=()):
+
+    """
+    Get the return type of a masked UDF for a given set of argument dtypes. It
+    is assumed that the function consumes a dictionary whose keys are strings
+    and whose values are of MaskedType. Initially assume that the UDF may be
+    written to utilize any field in the row - including those containing an
+    unsupported dtype. If an unsupported dtype is actually used in the function
+    the compilation should fail at `compile_udf`. If compilation succeeds, one
+    can infer that the function does not use any of the columns of unsupported
+    dtype - meaning we can drop them going forward and the UDF will still end
+    up getting fed rows containing all the fields it actually needs to use to
+    compute the answer for that row.
+    """
+
+    # present a row containing all fields to the UDF and try and compile
+    compile_sig = (argty, *(typeof(arg) for arg in args))
+
+    # Get the return type. The PTX is also returned by compile_udf, but is not
+    # needed here.
+    ptx, output_type = cudautils.compile_udf(func, compile_sig)
+    if not isinstance(output_type, MaskedType):
+        numba_output_type = numpy_support.from_dtype(np.dtype(output_type))
+    else:
+        numba_output_type = output_type
+
+    result = (
+        numba_output_type
+        if not isinstance(numba_output_type, MaskedType)
+        else numba_output_type.value_type
+    )
+
+    # _get_udf_return_type will throw a TypingError if the user tries to use
+    # a field in the row containing an unsupported dtype, except in the
+    # edge case where all the function does is return that element:
+
+    # def f(row):
+    #    return row[<bad dtype key>]
+    # In this case numba is happy to return MaskedType(<bad dtype key>)
+    # because it relies on not finding overloaded operators for types to raise
+    # the exception, so we have to explicitly check for that case.
+    if isinstance(result, Poison):
+        raise TypeError(str(result))
```

Comment:
Should we throw a numba TypingError for consistency?

Reply:
I actually agree with this, but would love @gmarkall thoughts. Usually `TypingError` is used to indicate something actually went wrong with the type inference stage inside Numba, like a compilation of a function was requested with a set of types that aren't compatible with what happens inside the function (like adding two datetimes). Normally that's actually what will happen when a user uses an unsupported dtype, this is the only tricky case that gets around that, because in these cases the UDF doesn't try and "use" any functions or operators so numba can't error by "not finding" them. So perhaps we should raise a `TypingError` after all?

Reply:
Agreed, curious what Graham thinks.

Reply:
Went with `TypingError` here